File: Development/TunerLib/src/com/mad/tunerlib/chordGeneration/FingerPosition.java
package com.mad.tunerlib.chordGeneration;

import com.mad.lib.chordGeneration.EString;
import com.mad.lib.musicalBase.Notes;
import com.mad.lib.musicalBase.OctavedNote;
import com.mad.lib.utils.ErrorHandler;
import com.mad.tunerlib.musicalBase.FretFingerPair;

public class FingerPosition
{
	public int		String;
	public int		Fret;

	static int[]	s_arFretDifference	= { 5, 5, 5, 4, 5,
			FretFingerPair.NUM_FRETS	};

	public static int sameNoteFretDifferenceOnString(	final int eCurrentString,
														final int eNextString)
	{

		int nJumpString, nMul = 1;
		if (eNextString < eCurrentString)
		{
			nJumpString = eNextString;
			nMul = -1;
		}
		else
		{
			nJumpString = eCurrentString;
		}

		if ((nJumpString < EString.Sixth)
				|| (nJumpString > EString.LastString))
		{
			Exception ex = new IndexOutOfBoundsException();
			ErrorHandler.HandleError(ex);
			return 0;
		}
		return s_arFretDifference[nJumpString] * nMul;

	}

	public static FingerPosition getFirstPosition()
	{
		return new FingerPosition(EString.Sixth, 0);

	}

	public FingerPosition(final FingerPosition other)
	{
		String = other.String;
		Fret = other.Fret;
	}

	public FingerPosition(final int eString, final int nFret)
	{
		String = eString;
		Fret = nFret;
	}

	public boolean advanceFrets(final int nNumberOfFrets)
	{
		if (nNumberOfFrets < 0)
		{
			ErrorHandler
					.HandleError(new IllegalStateException());
		}
		if (nNumberOfFrets == 0)
		{
			return true;
		}

		if (IsLastPosition())
		{
			return PreviousOctave();
		}

		int nNumberOfFretsToAdvance = nNumberOfFrets;

		while (nNumberOfFretsToAdvance != 0)
		{
			// If we are
			int nNextString =
					nNumberOfFretsToAdvance > 0 ? String + 1
							: String - 1;
			int nSameNotesDifference =
					sameNoteFretDifferenceOnString(String,
							nNextString);
			int nCurrFretsToAdvance =
					Math.min(Math.abs(nNumberOfFretsToAdvance),
							nSameNotesDifference);
			if (nSameNotesDifference > 0)
			{
				if (((Fret + nCurrFretsToAdvance) >= nSameNotesDifference)
						|| ((Fret + nCurrFretsToAdvance) < 0))
				{
					if ((nNextString < 0)
							|| (nNextString > EString.LastString))
					{
						break;
					}

					Fret =
							(Fret + nCurrFretsToAdvance + nSameNotesDifference)
									% nSameNotesDifference;
					String = nNextString;

				}
				else
				{
					Fret += nCurrFretsToAdvance;
				}
			}
			nNumberOfFretsToAdvance -= nCurrFretsToAdvance;
		}

		return true;
	}

	public boolean Equals(final FingerPosition other)
	{
		return (String == other.String) && (Fret == other.Fret);
	}

	public int getAbsouluteFretOffset()
	{
		int result = 0;
		for (int nGuitarString = EString.FirstString; nGuitarString < String; nGuitarString++)
		{
			result +=
					sameNoteFretDifferenceOnString(
							nGuitarString + 1, nGuitarString);
		}

		result += Fret;

		return result;
	}

	boolean isBefore(final FingerPosition other)
	{
		if (String == other.String)
		{
			return Fret < other.Fret;
		}

		if (String < other.String)
		{
			int nFretToCheck =
					other.Fret
							+ ((other.String - String) * Notes.NumberOfNotes);

			return Fret < nFretToCheck;
		}
		else
		{
			int nFretToCheck =
					Fret
							+ ((String - other.String) * Notes.NumberOfNotes);

			return nFretToCheck < other.Fret;
		}
	}

	boolean IsFirstPosition()
	{
		return this.Equals(FingerPosition.getFirstPosition());
	}

	boolean IsFirstPosition(final FingerPosition pos)
	{
		return pos.IsFirstPosition();
	}

	boolean isFretValid(final int nFret)
	{
		if ((nFret >= com.mad.lib.utils.Definitions.NUMBER_OF_FRETS)
				|| (nFret < 0))
		{
			return false;
		}

		return true;
	}

	boolean IsLastPosition()
	{
		return (Fret == com.mad.lib.utils.Definitions.NUMBER_OF_FRETS)
				&& (String == EString.NumberOfStrings);
	}

	boolean IsLastPosition(final FingerPosition pos)
	{
		return pos.IsLastPosition();
	}

	public boolean isValid()
	{
		if (!isFretValid(Fret))
		{
			return false;
		}

		if ((String < EString.FirstString)
				|| (String > EString.LastString))
		{
			return false;
		}

		return true;
	}

	public boolean NextOctave()
	{
		return advanceFrets(Notes.NumberOfNotes);
	}

	public boolean NextString()
	{
		if (String == EString.LastString)
		{
			return false;
		}

		int nFret =
				Fret
						+ sameNoteFretDifferenceOnString(String,
								String + 1);

		if (!isFretValid(nFret))
		{
			return false;
		}
		String++;
		Fret = nFret;

		return true;
	}

	public boolean PreviousOctave()
	{
		return advanceFrets(-1 * Notes.NumberOfNotes);
	}

	public boolean previousString()
	{
		if (String == EString.FirstString)
		{
			return false;
		}

		int nFret =
				Fret
						+ sameNoteFretDifferenceOnString(String,
								String - 1);

		if (!isFretValid(nFret))
		{
			return false;
		}

		Fret = nFret;
		String--;
		return true;
	}

	public OctavedNote getOctavedNote()
	{
		int fretsToAdd = 0;
		for (int i = 1; i < String; i++)
		{
			fretsToAdd +=
					FingerPosition
							.sameNoteFretDifferenceOnString(
									i - 1, i);
		}

		return new OctavedNote(fretsToAdd + Fret
				+ (Notes.E + (2 * Notes.NumberOfNotes)));
	}

};

File: Development/ChordFingerFinder/FingerPosition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ChordFingerFinder
{
    class CFingerPosition
    {

        public EString String { get; private set; }
        public int Fret { get; private set; }

        public CFingerPosition(CFingerPosition other)
        {
            String = other.String;
            Fret = other.Fret;
        }

        bool IsBefore(CFingerPosition other)
        {
            if (String == other.String)
            {
                return Fret < other.Fret;
            }

            if (String < other.String)
            {
                int nFretToCheck =
                    other.Fret + (other.String - String) * Globals.NUMBER_OF_FRETS_IN_OCTAVE;

                return Fret < nFretToCheck;
            }
            else
            {
                int nFretToCheck =
                    Fret + (String - other.String) * Globals.NUMBER_OF_FRETS_IN_OCTAVE;

                return nFretToCheck < other.Fret;
            }
        }

        public bool IsValid()
        {
            if (!IsFretValid(Fret))
            {
                return false;
            }

            if (String < EString.FirstString ||
                String > EString.LastString)
            {
                return false;
            }

            return true;
        }

        bool IsFretValid(int nFret)
        {
            if (nFret >= Globals.NUMBER_OF_FRETS || nFret < 0)
            {
                return false;
            }

            return true;
        }

        public bool PreviousString()
        {
            if (String == EString.FirstString)
            {
                return false;
            }

            int nFret = Fret + SameNoteFretDifferenceOnString(String, String - 1);

            if (!IsFretValid(nFret))
            {
                return false;
            }

            Fret = nFret;
            String--;
            return true;
        }

        public bool NextString()
        {
            if (String == EString.LastString)
            {
                return false;
            }

            int nFret = Fret + SameNoteFretDifferenceOnString(String, String + 1);

            if (!IsFretValid(nFret))
            {
                return false;
            }
            String++;
            Fret = nFret;

            return true;
        }

        public bool NextOctave()
        {
            return AdvanceFrets(Globals.NUMBER_OF_FRETS_IN_OCTAVE);
        }

        public bool PreviousOctave()
        {
            return AdvanceFrets(-1 * Globals.NUMBER_OF_FRETS_IN_OCTAVE);
        }

        public bool AdvanceFrets(int nNumberOfFrets)
        {
            if (nNumberOfFrets == 0)
            {
                return true;
            }

            if (IsLastPosition())
            {
                return PreviousOctave();
            }

            int nNumberOfFretsToAdvance = nNumberOfFrets;

            while (nNumberOfFretsToAdvance != 0)
            {
                int nSameNotesDifference = 0;

                if (nNumberOfFretsToAdvance < 0)
                {
                    nSameNotesDifference = SameNoteFretDifferenceOnString(String, String - 1);
                }
                else
                {
                    nSameNotesDifference = SameNoteFretDifferenceOnString(String, String + 1);
                }

                int nAbsSameNotesDifference = Math.Abs(nSameNotesDifference);

                if (Math.Abs(nNumberOfFretsToAdvance) >= nAbsSameNotesDifference)
                {
                    if (nNumberOfFretsToAdvance < 0)
                    {
                        String--;
                        nNumberOfFretsToAdvance += nAbsSameNotesDifference;
                        continue;
                        
                    }
                    else
                    {
                        if (String == EString.LastString)
                        {
                            return false;
                        }

                        String++;
                        nNumberOfFretsToAdvance -= nAbsSameNotesDifference;
                        continue;
                    }
                }

                if (nNumberOfFretsToAdvance < 0)
                {
                    // Check if the fret is the first one.
                    if (Fret == 0)
                    {
                        Fret += nSameNotesDifference;
                        String--;
                    }

                    Fret--;

                }
                else
                {

                    if (Fret == Globals.NUMBER_OF_FRETS)
                    {
                        Fret += nSameNotesDifference;
                        String++;
                    }

                    Fret++;
                }

                nNumberOfFretsToAdvance--;
            }

            return true;
        }

        public int SameNoteFretDifferenceOnString(EString eCurrent, EString eNext)
        {
            if (Math.Abs(eCurrent - eNext) != 1)
            {
                throw new Exception("Those strings are not close");
            }

            if ((eCurrent == EString.Second) && (eNext == EString.Third))
            {
                return 4;
            }

            if ((eCurrent == EString.Third) && (eNext == EString.Second))
            {
                return -4;
            }

            if (eCurrent < eNext)
            {
                return -5;
            }
            else
            {
                return 5;
            }
        }

        bool IsFirstPosition()
        {
            return this.Equals(Globals.s_posFirstGuitarPosition);
        }

        bool IsLastPosition()
        {
            return (Fret == Globals.NUMBER_OF_FRETS) &&
                   (String == EString.NumberOfStrings);
        }

        bool IsFirstPosition(CFingerPosition pos)
        {
            return pos.IsFirstPosition();
        }

        bool IsLastPosition(CFingerPosition pos)
        {
            return pos.IsLastPosition();
        }

        public CFingerPosition(EString eString, int nFret)
        {
            String = eString;
            Fret = nFret;
        }

        public bool Equals(CFingerPosition other)
        {
            return String == other.String &&
                    Fret == other.Fret;
        }
    };
}
